aten/src/ATen/native/UnaryOps.cpp

```diff
@@ -49,8 +79,17 @@ static inline Tensor& unary_op_impl_out(Tensor& result, const Tensor& self, Stub
 // For example it must be at::bitwise_not_out instead of bitwise_not_out(which is at::native!).
 template <typename OutImpl>
 static inline Tensor unary_op_impl(const Tensor& self, OutImpl& out_impl) {
-  Tensor result = at::empty({0}, self.options());
-  return out_impl(result, self);
+  // This enables int-to-float implicit dtype conversions
+  ScalarType promoted_dtype = promoteIntToFloats(self);
```

Comment:
we'll need a way to differentiate which ops shouldn't promote int->float, e.g. `torch.abs()`

Reply:
From the checks I did on the numpy ops, there are 4 types of conversions on the Unary Ops:

 1. Type - 1: int8 - float16, int16 - float32, int32 - float64, int64 - float64, bool - float16
 2. Type - 2: (int8, int16, int32, int64, bool) - float64
 3. Type - 3: bool - int8
 4. Type - 4: bool - float16

Also, some ops are not defined in NumPy (like `sigmoid`, `polygamma`, `mvlgamma`). What do you suggest? Should we follow the type promotion similar to NumPy's strategy? Which, in case, will lead us to have conditions for the ops to go through a specific type (1/2/3/4) of dtype promotion.

In case you are interested to know which ops use what type of promotion:

 * Ops with no upcasting: `abs`, `real`, `imag`, `bitwise_not`, `clip`, `neg`
 * Ops with changes (Type 1): `ceil`, `expm1`, `floor`, `log`, `log10`, `log1p`, `log2`, `sin`, `sinh`, `sqrt`, `trunc`, `atan`, `cos`, `tan`
 * Ops with changes (Type 2): `angle`
 * Ops with changes (Type 3): `conjugate`, `reciprocal`, `square`
 * Ops with changes (Type 4): `round`

Reply:
@gchanan do you think we'd be willing to deviate from NumPy on some of these UnaryOp promotions?

0. (not promoting) these seem reasonable.
1. Seems reasonable.
2. This case for one operator (angle) is expecting a complex argument. The angle of any real argument is 0. These ops currently handle (most) integral types without promoting, and I'm not convinced we should change that behavior as part of this ticket.
3. `bool -> int8` for conjugate/reciprocal/square seem a bit weird to me. E.g. I'd expect `t.square()` to be the same as `t * t`. It's also possible we'd want to skip boolean tensors for these operators, since ops like bool_tensor.reciprocal() would crash if they contain any False values.
4. `round` This inconsistent behavior seems odd since none of the other integral types promote for that op. I'd be tempted to deviate from np on this one, or just not implement for bool.